dds/DCPS/XTypes/DynamicDataImpl.cpp

```diff
@@ -7,3263 +7,6479 @@
+  }
+  CORBA::ULong length = 1;
+  const DDS::BoundSeq& bound = descriptor->bound();
+  for (CORBA::ULong i = 0; i < bound.length(); ++i) {
+    length *= bound[i];
+  }
+
+  const DDS::DynamicType_var elem_type = get_base_type(descriptor->element_type());
+  const TypeKind elem_tk = elem_type->get_kind();
+  DDS::TypeDescriptor_var elem_td;
+  if (elem_type->get_descriptor(elem_td) != DDS::RETCODE_OK) {
+    return false;
+  }
+
+  if (data_->is_basic_type(elem_tk)) {
+    serialized_size_primitive_array(encoding, size, elem_tk, length);
+    return true;
+  } else if (elem_tk == TK_STRING8) {
+    IndexToIdMap index_to_id(length, MEMBER_ID_INVALID);
+    if (!get_index_to_id_map(index_to_id, length)) {
+      return false;
+    }
+    return serialized_size_generic_string_array<const char*>(encoding, size, index_to_id);
+  } else if (elem_tk == TK_STRING16) {
+#ifdef DDS_HAS_WCHAR
+    IndexToIdMap index_to_id(length, MEMBER_ID_INVALID);
+    if (!get_index_to_id_map(index_to_id, length)) {
+      return false;
+    }
+    return serialized_size_generic_string_array<const CORBA::WChar*>(encoding, size, index_to_id);
+#else
+    return false;
+#endif
+  } else if (elem_tk == TK_ENUM) {
+    const CORBA::ULong bit_bound = elem_td->bound()[0];
+    serialized_size_enum_array(encoding, size, length, bit_bound);
+    return true;
+  } else if (elem_tk == TK_BITMASK) {
+    const CORBA::ULong bit_bound = elem_td->bound()[0];
+    serialized_size_bitmask_array(encoding, size, length, bit_bound);
+    return true;
+  } else if (elem_tk == TK_SEQUENCE) {
+    const DDS::DynamicType_var nested_elem_type = get_base_type(elem_td->element_type());
+    const TypeKind nested_elem_tk = nested_elem_type->get_kind();
+    IndexToIdMap index_to_id(length, MEMBER_ID_INVALID);
+    if (!get_index_to_id_map(index_to_id, length)) {
+      return false;
+    }
+    if (data_->is_basic_type(nested_elem_tk)) {
+      return serialized_size_nesting_basic_array(encoding, size, nested_elem_tk, index_to_id);
+    } else if (nested_elem_tk == TK_ENUM) {
+      return serialized_size_nesting_enum_array(encoding, size, index_to_id);
+    } else if (nested_elem_tk == TK_BITMASK) {
+      return serialized_size_nesting_bitmask_array(encoding, size, index_to_id);
+    }
+  }
+
+  // Elements stored in complex map
+  IndexToIdMap index_to_id;
+  if (!get_index_to_id_from_complex(index_to_id, length)) {
+    return false;
+  }
+  return serialized_size_complex_array(encoding, size, index_to_id, elem_type);
+}
+
+bool DynamicDataImpl::DataContainer::serialize_array(DCPS::Serializer& ser) const
+{
+  DDS::TypeDescriptor_var descriptor;
+  if (type_->get_descriptor(descriptor) != DDS::RETCODE_OK) {
+    return false;
+  }
+  CORBA::ULong length = 1;
+  const DDS::BoundSeq& bound = descriptor->bound();
+  for (CORBA::ULong i = 0; i < bound.length(); ++i) {
+    length *= bound[i];
+  }
+
+  const DDS::DynamicType_var elem_type = get_base_type(descriptor->element_type());
+
```

Comment:
This should be able to use the same functions as XCDR2 like we do in normal serialization. Functions like `serialized_size_parameter_id` were designed to be able to handle either encoding, but I understand not covering it.

Reply:
Yes, I think most of these functions can be reused for XCDR1. The main difference is optional members for which I think we don't have an existing function that handles both encodings.